src/VarArray.mo

```diff
@@ -3,46 +3,162 @@
 import Iter "type/Iter";
 import Order "Order";
 import Result "Result";
+import Option "Option";
 import Prim "mo:⛔";
 import { todo } "Debug";
 
 module {
 
+  /// Creates an empty mutable array (equivalent to `[var]`).
   public func empty<T>() : [var T] = [var];
 
-  public func init<T>(size : Nat, initValue : T) : [var T] = Prim.Array_init<T>(size, initValue);
+  /// Creates a mutable array containing `item` repeated `size` times.
+  ///
+  /// ```motoko include=import
+  /// let array = VarArray.repeat<Nat>("Echo", 3);
+  /// assert array == [var "Echo", "Echo", "Echo"];
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  public func repeat<T>(item : T, size : Nat) : [var T] = Prim.Array_init<T>(size, item);
+
+  /// Duplicates `array`, returning a shallow copy of the original.
+  ///
+  /// ```motoko include=import
+  /// let array1 = [var 1, 2, 3];
+  /// let array2 = VarArray.clone<Nat>(array1);
+  /// array2[0] := 0;
+  /// assert array1 == [var 1, 2, 3];
+  /// assert array2 = [var 0, 2, 3];
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  public func clone<T>(array : [var T]) : [var T] {
+    let size = array.size();
+    if (size == 0) {
+      return [var]
+    };
+    let newArray = Prim.Array_init<T>(size, array[0]);
+    var i = 0;
+    while (i < size) {
+      newArray[i] := array[i];
+      i += 1
+    };
+    newArray
+  };
 
-  public func generate<T>(size : Nat, generator : Nat -> T) : [var T] {
+  /// Creates an immutable array of size `size`. Each element at index i
+  /// is created by applying `generator` to i.
+  ///
+  /// ```motoko include=import
+  /// let array : [var Nat] = VarArray.tabulate<Nat>(4, func i = i * 2);
+  /// assert array == [var 0, 2, 4, 6];
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `generator` runs in O(1) time and space.
+  public func tabulate<T>(size : Nat, generator : Nat -> T) : [var T] {
```

Comment:
Note to self: We should implement this as a runtime primitive, like Array_tabulate. It would be a lot more efficient (no bounds checks etc)

Reply:
Yes this would be good, since it's probably the most frequently-used function in this module for the base library implementation. Perhaps `VarArray_tabulate()`?